benchmark/utils/generator.hpp

```diff
@@ -0,0 +1,278 @@
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GINKGO_BENCHMARK_UTILS_GENERATOR_HPP
+#define GINKGO_BENCHMARK_UTILS_GENERATOR_HPP
+
+
+#include <string>
+
+
+#include "benchmark/utils/formats.hpp"
+#include "benchmark/utils/general.hpp"
+#include "benchmark/utils/loggers.hpp"
+#include "benchmark/utils/stencil_matrix.hpp"
+#if GINKGO_BUILD_MPI
+#include "benchmark/utils/distributed_helpers.hpp"
+#endif
+
+
+template <typename ValueType = etype, typename IndexType = itype>
+struct DefaultSystemGenerator {
+    using index_type = IndexType;
+    using value_type = ValueType;
+    using Vec = vec<ValueType>;
+
+    static gko::matrix_data<ValueType, IndexType> generate_matrix_data(
+        rapidjson::Value& config)
+    {
+        if (config.HasMember("filename")) {
+            std::ifstream in(config["filename"].GetString());
+            return gko::read_generic_raw<ValueType, IndexType>(in);
+        } else if (config.HasMember("stencil")) {
+            return generate_stencil<ValueType, IndexType>(
+                config["stencil"].GetString(), config["size"].GetInt64());
+        } else {
+            throw std::runtime_error(
+                "No known way to generate matrix data found.");
+        }
+    }
+
+    static std::shared_ptr<gko::LinOp> generate_matrix(
+        std::shared_ptr<const gko::Executor> exec, rapidjson::Value& config)
+    {
+        auto data = generate_matrix_data(config);
+        return generate_matrix_with_format(
+            std::move(exec), config["optimal"]["spmv"].GetString(), data);
+    }
+
+    static std::shared_ptr<gko::LinOp> generate_matrix_with_format(
+        std::shared_ptr<const gko::Executor> exec,
+        const std::string& format_name,
+        const gko::matrix_data<ValueType, itype>& data)
+    {
+        return gko::share(
+            ::formats::matrix_factory(format_name, std::move(exec), data));
+    }
+
+    static std::shared_ptr<gko::LinOp> generate_matrix_with_default_format(
+        std::shared_ptr<const gko::Executor> exec,
+        const gko::matrix_data<ValueType, itype>& data)
+    {
+        return generate_matrix_with_format(std::move(exec), "coo", data);
```

Comment:
Should coo stay the default? I think csr might be another suitable candidate.

Reply:
I didn't want to change any existing behavior, so I just kept the currently used default format. I don't think I would want to mix that discussion in here.